weblab/entities/tests/test_views.py

```diff
@@ -168,41 +237,106 @@ def test_create_protocol_version(self, user, client):
         assert protocol.repo.head.commit.author.name == user.full_name
 
     def test_create_protocol_version_requires_permissions(self, user, client):
-        model = ProtocolEntity.objects.create(
-            name='myprotocol',
-            visibility='public',
-            author=user,
-        )
+        protocol = recipes.protocol.make(author=user)
         response = client.post(
-            '/entities/protocols/' + str(model.pk) + '/versions/new',
+            '/entities/protocols/%d/versions/new' % protocol.pk,
             data={},
         )
         assert response.status_code == 302
         assert '/login/' in response.url
 
 
 @pytest.mark.django_db
-def test_upload_file(user, client):
-    model = ModelEntity.objects.create(
-        name='mymodel',
-        visibility='public',
-        author=user,
-    )
+class TestFileUpload:
+    def test_upload_file(self, user, client):
+        model = recipes.model.make(author=user)
 
-    upload = io.StringIO('my test model')
-    upload.name = 'model.txt'
-    response = client.post(
-        '/entities/' + str(model.pk) + '/upload-file',
-        {
-            'upload': upload
-        }
-    )
+        upload = io.StringIO('my test model')
+        upload.name = 'model.txt'
+        response = client.post(
+            '/entities/%s/upload-file' % model.pk,
+            {
+                'upload': upload
+            }
+        )
+
+        data = json.loads(response.content.decode())
+        upload = data['files'][0]
+        assert upload['stored_name'] == 'uploads/model.txt'
+        assert upload['name'] == 'model.txt'
+        assert upload['is_valid']
+        assert upload['size'] == 13
+
+        assert model.files.count() == 1
+
+    def test_bad_upload(self, user, client):
+        model = recipes.model.make(author=user)
+
+        response = client.post('/entities/%s/upload-file' % model.pk, {})
+
+        assert response.status_code == 400
+
+
+@pytest.mark.django_db
+class TestEntityVisibility:
+    def test_private_entity_visible_to_self(self, client, user):
+        model = recipes.model.make(visibility='private', author=user)
+        add_version(model)
+        assert client.get('/entities/models/%s' % model.pk, follow=True).status_code == 200
+
+    def test_private_entity_invisible_to_other_user(self, client, user, other_user):
+        model = recipes.model.make(visibility='private', author=other_user)
+        add_version(model)
+        response = client.get('/entities/models/%s' % model.pk)
+        assert response.status_code == 302
+        assert '/login' in response.url
```

Comment:
While redirecting to the login page is the right thing where the user is anonymous, I'm not sure it's so logical if the user is logged in but doesn't have permissions. In that case wouldn't a 403 or 404 response be more appropriate (depending on whether we want people to know the ID exists)?

Reply:
Good call, I agree. Anonymous user should see entity if they have access, otherwise login redirect (regardless of whether it exists). Logged in user should see entity if they have access, otherwise 404.